dolphin/alert_manager/trap_receiver.py

```diff
@@ -0,0 +1,137 @@
+
+LOG = log.getLogger(__name__)
+
+#Currently one mib file is loaded, All files to be loaded
+LOAD_MIB_MODULE = 'SNMPv2-MIB'
+
+class TrapReceiver(manager.Manager):
+    """Trap receiver functions"""
+
+    def __init__(self,receiverIpAddr=None, *args, **kwargs):
+        super(TrapReceiver, self).__init__(*args, **kwargs)
+
+    def _mib_builder(self, load_mibs_list):
+        mibBuilder = builder.MibBuilder()
+        try:
+            global mibViewController
+            mibViewController = view.MibViewController(mibBuilder)
+            if load_mibs_list:
+                _mibs = LOAD_MIB_MODULE.split(",")
+                mibBuilder.loadModules(*_mibs)
+        except error.MibNotFoundError as excep:
+            LOG.info(" {} Mib Not Found!".format(excep))
+
+    def _add_transport(self):
+        """
+        :return:
+        """
+        try:
+            config.addTransport(
+                self.snmpEngine,
+                udp.domainName,
+                udp.UdpTransport().openServerMode((constants.DEF_TRAP_RECV_ADDR, int(constants.DEF_TRAP_RECV_PORT)))
+            )
+        except Exception as e:
+            LOG.info("{} Port Binding Failed the Provided Port {} is in Use".format(e, self.receiverPort))
+
+    def _cbFun(self, stateReference, contextEngineId, contextName,
+              varBinds, cbCtx):
+        global mibViewController
+        LOG.info("####################### NEW Notification #######################")
+        execContext = self.snmpEngine.observer.getExecutionContext('rfc3412.receiveMessage:request')
+        LOG.info(
+            '#Notification from %s \n#ContextEngineId: "%s" \n#ContextName: "%s" \n#SNMPVER "%s" \n#SecurityName "%s"' % (
+            '@'.join([str(x) for x in execContext['transportAddress']]), contextEngineId.prettyPrint(),
+            contextName.prettyPrint(), execContext['securityModel'], execContext['securityName']))
+        for oid, val in varBinds:
+            output = rfc1902.ObjectType(rfc1902.ObjectIdentity(oid), val).resolveWithMib(
+                mibViewController).prettyPrint()
+            LOG.info(output)
+
+    def _configure_snmp_userpara(self):
+        """
+        :return:
+        """
+
+        COMMUNITYSTRING = constants.SNMP_COMMUNITY_STR
+        config.addV1System(self.snmpEngine, COMMUNITYSTRING, COMMUNITYSTRING)
+        __authProtocol = {
+            'usmHMACMD5AuthProtocol': config.usmHMACMD5AuthProtocol,
+            'usmHMACSHAAuthProtocol': config.usmHMACSHAAuthProtocol,
+            'usmAesCfb128Protocol': config.usmAesCfb128Protocol,
+            'usmAesCfb256Protocol': config.usmAesCfb256Protocol,
+            'usmAesCfb192Protocol': config.usmAesCfb192Protocol,
+            'usmDESPrivProtocol': config.usmDESPrivProtocol,
+            'usmNoAuthProtocol': config.usmNoAuthProtocol,
+            'usmNoPrivProtocol': config.usmNoPrivProtocol
+        }
+
+
+        config.addV3User(
+            self.snmpEngine, userName=constants.SNMP_USM_USER,
```

Comment:
This user info is the USM user info hat is going to be sending the trap.? How do we achieve hetrogenioty here? Do we start one trap reciever instance for each device registered?

Reply:
For the usm user,there is remote engine id which is unique for device. For v2, there is a way to configure comunity string along with device ip. Will update code for V2 part

Reply:
My question is, is config.addV3User is  to add a backend device SNMP user info? if so how do we map different device user info ?

Reply:
config.addV3User is to add snmp user info to snmp engine(not to backened). Here prerequisite is snmp user info at device side is configured. remote engine id  field uniquely identifies the device which sends V3 trap. so snmp engine can differentiate between various devices and their corresponding user configurations